src/Razor/test/Microsoft.VisualStudio.Razor.Integration.Test/InProcess/EditorInProcess_Classification.cs

```diff
@@ -0,0 +1,141 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Classification;
+using Microsoft.VisualStudio.Text.Editor;
+using Xunit;
+
+namespace Microsoft.VisualStudio.Extensibility.Testing
+{
+    internal partial class EditorInProcess
+    {
+        /// <summary>
+        /// Waits for any semantic classifications to be available on the active TextView, and for at least one of the
+        /// <paramref name="expectedClassification"/> if provided.
+        /// </summary>
+        /// <param name="cancellationToken">A cancellation token.</param>
+        /// <param name="expectedClassification">The classification to wait for, if any.</param>
+        /// <param name="count">The number of the given classification to expect.</param>
+        /// <returns>A <see cref="Task"/> which completes when classification is "ready".</returns>
+        public async Task WaitForClassificationAsync(CancellationToken cancellationToken, string expectedClassification = "RazorComponentElement", int count = 1)
+        {
+            var textView = await TestServices.Editor.GetActiveTextViewAsync(cancellationToken);
+            var classifier = await GetClassifierAsync(textView, cancellationToken);
+
+            using var semaphore = new SemaphoreSlim(1);
+            await semaphore.WaitAsync(cancellationToken);
+
+            classifier.ClassificationChanged += Classifier_ClassificationChanged;
+
+            // Check that we're not ALREADY changed
+            if (await HasClassificationAsync(cancellationToken))
+            {
+                semaphore.Release();
+                classifier.ClassificationChanged -= Classifier_ClassificationChanged;
+                return;
+            }
+
+            try
+            {
+                await semaphore.WaitAsync(cancellationToken);
+            }
+            finally
+            {
+                classifier.ClassificationChanged -= Classifier_ClassificationChanged;
+            }
+
+            async void Classifier_ClassificationChanged(object sender, ClassificationChangedEventArgs e)
```

Comment:
`async void` is _always_ the wrong choice (even in cases where it works, it's unnecessarily more complex than alternatives). Consider enabling the vs-threading analyzer to disallow this and it will prevent problems from sneaking through.

Reply:
Good catch, didn't realize I'd done that, I refactored to avoid it and filed https://github.com/dotnet/razor-tooling/issues/5939 to disallow that by policy.